### ISSUE TITLE
Can't parse 'between' statement when expression starts with minus sign

### ISSUE THREAD

**Issue by MarcAntoineSchmidtQC** [sqlfluff/sqlfluff#2845]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff is unable to parse a between statement when one of the expression starts with an arithmetic operator (e.g. minus sign)

I think a fix could be to change the lines here: 

https://github.com/sqlfluff/sqlfluff/blob/e5e5c7e8f737f391bc962722d83ab6bb5367911c/src/sqlfluff/dialects/dialect_ansi.py#L1659-L1665

to       
```python
    Ref("ArithmeticBinaryOperatorGrammar", optional=True)
    Ref("Expression_C_Grammar"),
    AnyNumberOf(
        Sequence(
            Ref("ArithmeticBinaryOperatorGrammar"),
            Ref("Expression_C_Grammar"),
        )
    ),
```

Related: issue #478 and PR #498. 

### Expected Behaviour

I expect the parsing to succeed

### Observed Behaviour

The parsing fails

### How to reproduce

Trying to parse the following file will result in an error:

```sql
SELECT CASE WHEN a BETWEEN -b AND 0 THEN 1 ELSE 0 END
FROM my_table
```

but this will succeed:
```sql
SELECT CASE WHEN a BETWEEN b AND 0 THEN 1 ELSE 0 END
FROM my_table
```

### Dialect

TSQL

### Version

sqlfluff, version 0.11.0
Python 3.7.12

### Configuration

```
[sqlfluff]
nocolor = True
dialect = tsql
templater = jinja
exclude_rules = L027,L031,L059

[sqlfluff:rules]
comma_style = leading
max_line_length = 120

[sqlfluff:rules:L016]
ignore_comment_lines = True
ignore_comment_clauses = True
```

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
Good skills. I'd suggest a slight tweak instead to this:

```py
    Sequence(
        Ref("ArithmeticBinaryOperatorGrammar", optional=True),
        Ref("Expression_C_Grammar"),
        allow_gaps=False,
    ),
    AnyNumberOf(
        Sequence(
            Ref("ArithmeticBinaryOperatorGrammar"),
            Ref("Expression_C_Grammar"),
        )
    ),
    "AND",
    Sequence(
        Ref("ArithmeticBinaryOperatorGrammar", optional=True),
        Ref("Expression_C_Grammar"),
        allow_gaps=False,
    ),
```

Otherwise it would parse `...WHEN a BETWEEN - b AND 0 THEN...` (with a gap) when it probably shouldn't.

Also the final one will allow it to parse `...WHEN a BETWEEN - b AND -c THEN...`

Fancy opening a PR with this change and a couple of test cases? More details here if you need them: https://github.com/sqlfluff/sqlfluff/wiki/Contributing-Dialect-Changes

**Comment by MarcAntoineSchmidtQC** [comment]
I'll open a PR!

Regarding the gap, I'm not sure I understand. I don't know about ANSI in general but with TSQL, having a gap between the minus sign and the expression still works. Agreed it's not nice to read but it parses. Why would we want sqlfluff not to parse this expression correctly?

**Comment by tunetheweb** [comment]
> Why would we want sqlfluff not to parse this expression correctly?

Cause I (incorrectly) presumed it wasn't valid 😁

Using `ArithmeticBinaryOperatorGrammar` (which includes `MinusSegment` rather than `NegativeSegment`) will mean that L006 _insists_ on a whitespace, which is probably not what you want.

It would be better to use `NegativeSegment` instead. This will also prevent invalid values like `...WHEN a BETWEEN *b AND 0 THEN...`

In fact, in that case, it might be better to add `NegativeSegment` to start of `Expression_D_Grammar` instead?:

```py
    # Expression_D_Grammar
    # https://www.cockroachlabs.com/docs/v20.2/sql-grammar.htm#d_expr
    Expression_D_Grammar=Sequence(
        Ref("NegativeSegment", optional=True),
        OneOf(
        ...
```
